### DISCUSSION TITLE
Using Dagre For layouting - How to apply layout on first render

### DISCUSSION THREAD

**Discussion by royhuntley** [xyflow/xyflow#4167] [Layouting]
Am just starting to use Dagre as a layouting library.  Using the example here : https://reactflow.dev/learn/layouting/layouting
I can see the layouting works when clicking either horizontal or vertical.  How do i get the layout to apply on the first render of the nodes (ie, without having to click the button) first. Thanks

**Comment by latifs** [comment]
@royhuntley any progress on this,
I'm also trying to re-layout when the data changes but the layout is all messed up for me.

Here's my issue:
https://github.com/xyflow/xyflow/discussions/4184

**Reply by royhuntley** [reply]
see below comment - ive found a way to apply the layout on first load.  
For your issue around there being no real hierarchy when using dagre, i think you need to apply 'rank' to certain nodes, but im afraid ive not looked in to this.

**Comment by royhuntley** [comment]
I have manage to solve my issue, of wanting to apply the layout upon first render.

const onLoad = useCallback( () => { onLayout('LR'); setTimeout(() => fitView(), 0) } );

and then...

<ReactFlow onLoad={onLoad} etc,,,,>
